lib/embulk/input/jira_api/client.rb

```diff
@@ -7,31 +7,83 @@ module Embulk
   module Input
     module JiraApi
       class Client
-        PARALLEL_THREAD_COUNT = 50
-        SEARCH_TIMEOUT_SECONDS = 5
-        SEARCH_ISSUES_TIMEOUT_SECONDS = 60
+        MAX_CONCURRENT_REQUESTS = 50
+        # Normal http request timeout is 300s
+        SEARCH_ISSUES_TIMEOUT_SECONDS = 300
         DEFAULT_SEARCH_RETRY_TIMES = 10
 
         def self.setup(&block)
           Jiralicious.configure(&block)
           new
         end
 
+        def find_at_key(array, key)
+          
+        end
+
         def search_issues(jql, options={})
-          timeout_and_retry(SEARCH_ISSUES_TIMEOUT_SECONDS) do
+          parallel_threads = MAX_CONCURRENT_REQUESTS
+          timeout_and_retry(SEARCH_ISSUES_TIMEOUT_SECONDS * MAX_CONCURRENT_REQUESTS ) do
             issues_raw = search(jql, options).issues_raw
-
             # TODO: below code has race-conditon.
-            Parallel.map(issues_raw, in_threads: PARALLEL_THREAD_COUNT) do |issue_raw|
-              # https://github.com/dorack/jiralicious/blob/v0.4.0/lib/jiralicious/search_result.rb#L32-34
-              issue = Jiralicious::Issue.find(issue_raw["key"])
-              JiraApi::Issue.new(issue)
+            success_items = []
+            fail_items = []
+            errors = []
+            search_issue_count = 0
+            
+            while issues_raw.length > 0 && search_issue_count <= DEFAULT_SEARCH_RETRY_TIMES do
+              search_results = Parallel.map(issues_raw, in_threads: parallel_threads) do |issue_raw|
+                # https://github.com/dorack/jiralicious/blob/v0.4.0/lib/jiralicious/search_result.rb#L32-34
+                begin
+                  issue = Jiralicious::Issue.find(issue_raw["key"])
+                  {
+                    :error => nil,
+                    :result => JiraApi::Issue.new(issue)
+                  }
+                rescue MultiJson::ParseError => e
+                  html = e.message
+                  title = html[%r|<title>(.*?)</title>|, 1] #=> e.g. "Unauthorized (401)"
+                  raise title if title != "Unauthorized (401)"
+                  {
+                    :error => e,
+                    :issue_raw => issue_raw,
+                    :result => nil
+                  }
+                end
+              end
+              search_issue_count += 1
+              for issue_result in search_results do
+                if !issue_result[:error].nil? 
+                  fail_items.push(issue_result[:issue_raw])
+                  errors.push(issue_result[:error])
+                else
+                  success_items.push(issue_result[:result])
+                end
+              end
+              raise errors[0] if search_issue_count > DEFAULT_SEARCH_RETRY_TIMES && errors.length > 0
```

Comment:
Since we only use first element, can we not store all errors?

Reply:
I changed it to a single variable

Reply:
Done